### PR TITLE
Feature/enable inference optimized qwen moe

### PR DESCRIPTION
**Pull request by utkarsh530**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/inference/moe/vllm_fused_moe.py [hunk 1]
```diff
@@ -637,8 +646,19 @@ def vllm_fused_moe(
         top_k=topk,
         config=config,
         grid_size=grid_size_fc1,
-        fuse_squared_relu=True,
+        fuse_squared_relu=not is_swiglu,
     )
+    if is_swiglu:
+        # gate = first half, up = second half (megatron chunk convention).
+        # Device-bounded Triton kernel instead of torch silu*mul: the torch version
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/moe/vllm_fused_moe.py]: This comment reads like a debugging log from Claude. Can we rewrite it?

File: megatron/core/transformer/moe/token_dispatcher_inference.py [hunk 2]
```diff
@@ -196,15 +196,26 @@ def token_dispatch(self, hidden_states, probs):
...
             with torch.no_grad():
-                self.routing_map = gather_from_sequence_parallel_region(
-                    self.routing_map, group=self.tp_ep_group
+                rmap = self.routing_map
+                h8 = hidden_states.contiguous().view(torch.uint8)
+                p8 = probs.contiguous().view(torch.uint8)
+                r8 = rmap.contiguous().view(torch.uint8)
+                nh, np_ = h8.shape[1], p8.shape[1]
+                packed = torch.cat((h8, p8, r8), dim=1)
+                gathered = gather_from_sequence_parallel_region(
```
Comment (sidsingh-nvidia) [inline megatron/core/transformer/moe/token_dispatcher_inference.py]: This is a good optimization, but probably belongs in a different PR?

**Review by sidsingh-nvidia — CHANGES_REQUESTED** [review]

File: megatron/core/inference/moe/vllm_fused_moe.py (hunk 1 above)
Comment (utkarsh530) [reply megatron/core/inference/moe/vllm_fused_moe.py]: Thanks for the review! Yeah, I have to clean up some things here; that's why I made it a draft

File: megatron/core/transformer/moe/token_dispatcher_inference.py (hunk 2 above)
Comment (utkarsh530) [reply megatron/core/transformer/moe/token_dispatcher_inference.py]: Yes! I just removed these paths

File: megatron/core/inference/moe/vllm_fused_moe.py [hunk 3]
```diff
@@ -637,8 +645,19 @@ def vllm_fused_moe(
         top_k=topk,
         config=config,
         grid_size=grid_size_fc1,
-        fuse_squared_relu=True,
+        fuse_squared_relu=not is_swiglu,
```
Comment (wdykas) [inline megatron/core/inference/moe/vllm_fused_moe.py:645]: If we are going to be having multiple activation functions should we make this arg more generic?

File: megatron/core/transformer/transformer_config.py [hunk 4]
```diff
@@ -1431,10 +1431,19 @@ def __post_init__(self):
                     "to avoid costly dtype conversions during decode."
                 )
 
-            if self.gated_linear_unit:
+            # Gated linear units (SwiGLU/GeGLU) are supported by the flashinfer, torch
+            # (mcore padded_swiglu), and vllm grouped-GEMM backends; only the remaining
```
Comment (sidsingh-nvidia) [inline megatron/core/transformer/transformer_config.py]: Aren't these all of the backends now? Is there a backend that will not support swiglu?

**Review by sidsingh-nvidia — CHANGES_REQUESTED** [review]

**Comment by sidsingh-nvidia** [comment]
The docstrings in this PR are very verbose. They read like debugging logs from Claude. Can we trim them and make them more readable?

**Review by sidsingh-nvidia — APPROVED** [review]
Approved pending docstring cleanup. Thanks for the contribution!

File: megatron/core/transformer/transformer_config.py (hunk 4 above)
Comment (utkarsh530) [reply megatron/core/transformer/transformer_config.py]: I think Flashinfer may not work and hasn't been fully tested on my side.

File: megatron/core/inference/moe/vllm_fused_moe.py (hunk 3 above)
Comment (utkarsh530) [reply megatron/core/inference/moe/vllm_fused_moe.py:645]: I am not sure about that, but I can help to generalize this, maybe in the next PR?

File: megatron/core/transformer/transformer_config.py [hunk 5]
```diff
@@ -1431,10 +1431,18 @@ def __post_init__(self):
                     "to avoid costly dtype conversions during decode."
                 )
 
-            if self.gated_linear_unit:
+            # Gated linear units are supported by the flashinfer, torch, and vllm
+            # grouped-GEMM backends; other backends still reject them.
+            _gated_ok_backends = ("flashinfer", "torch", "vllm")
+            _backend = (
+                str(getattr(self, "inference_grouped_gemm_backend", "")).lower().split(".")[-1]
```
Comment (santhnm2) [inline megatron/core/transformer/transformer_config.py]: Why do we need the `lower` and `split` calls here? This should already be one of `'flashinfer', 'torch', 'vllm'` right?

**Review by santhnm2 — APPROVED** [review]

File: megatron/core/transformer/transformer_config.py (hunk 5 above)
Comment (utkarsh530) [reply megatron/core/transformer/transformer_config.py]: Yeah thanks for catching this, I realized we don't need it

**Comment by utkarsh530** [comment]
I tested Flashinfer; it ran but gave garbage output. I am marking that right now it is not supported yet

**Review by yaox12 — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Review by shanmugamr1992 — APPROVED** [review]

**Comment by mathemakitten** [comment]
/ok to test [25156c7](https://github.com/NVIDIA/Megatron-LM/pull/5700/commits/25156c75f48a7fdd797500282204bb05d03bfa77)

**Comment by mathemakitten** [comment]
/ok to test [465d634](https://github.com/NVIDIA/Megatron-LM/pull/5700/commits/465d63464b2883c4d5b424b8299efda3e4279602)

**Review by chtruong814 — APPROVED** [review]

**Review by ko3n1g — APPROVED** [review]

**Review by balasaajay — APPROVED** [review]

**Review by jiemingz — APPROVED** [review]

File: megatron/training/models/gpt.py [hunk 6]
```diff
@@ -52,6 +52,14 @@ def default_layer_spec(config: "GPTModelConfig", vp_stage: int) -> ModuleSpec:
     transformer_cfg = config.transformer
     use_te = transformer_cfg.transformer_impl == "transformer_engine"
     if transformer_cfg.transformer_impl == "inference_optimized":
+        if transformer_cfg.num_moe_experts is not None:
```
Comment (maanug-nv) [inline megatron/training/models/gpt.py]: this branch already exists below. why are we duplicating that, instead of fixing the ordering? it seems like the moe experts branch should take priority over inference optimized branch? so either inference optimized branch should be moved down or moe experts branch should be moved up.

Comment (shanmugamr1992) [reply megatron/training/models/gpt.py]: Good catch — agreed, this was duplicating the shared `num_moe_experts` branch. Fixed by narrowing the `inference_optimized` branch to the dense (non-MoE) case and letting MoE models fall through to the shared MoE branch. `get_gpt_decoder_block_spec` already handles the `inference_optimized` impl internally, and `use_transformer_engine` resolves to `False` for `inference_optimized`, so behavior is unchanged.

Pushed as `shanmugamr1992:fix/dedupe-inference-optimized-moe-layer-spec` (commit 5e3277c15); will fold it into this PR branch.

**Comment by shanmugamr1992** [comment]
/ok to test [7dfaa4e](https://github.com/NVIDIA/Megatron-LM/pull/5700/commits/7dfaa4e95c48769d448b961667860f07d3126f01)

**Review by cuichenx — APPROVED** [review]

File: megatron/training/models/gpt.py (hunk 6 above)
Comment (maanug-nv) [reply megatron/training/models/gpt.py]: nice, that seems like the best way to handle it. thanks for addressing.

**Review by maanug-nv — APPROVED** [review]

**Comment by guihong-nv** [comment]
/ok to test [c75b9f6](https://github.com/NVIDIA/Megatron-LM/pull/5700/commits/c75b9f68cfaa896376c8d1759e4daad95a1aa4bc)

**Comment by mathemakitten** [comment]
/ok to test [63514ae](https://github.com/NVIDIA/Megatron-LM/pull/5700/commits/63514ae54acb7a1393d9876580df246dcf864423)

**Comment by guihong-nv** [comment]
/ok to test [de5b205](https://github.com/NVIDIA/Megatron-LM/pull/5700/commits/de5b2054f57e7dce6e28270f926ca69117e68890)

**Review by guihong-nv — APPROVED** [review]

**Comment by guihong-nv** [comment]
/ok to test [6bcdf4a](https://github.com/NVIDIA/Megatron-LM/pull/5700/commits/6bcdf4a192edac0c5a22e93f29e650da026675c1)

**Comment by guihong-nv** [comment]
/ok to test [ce43b4f](https://github.com/NVIDIA/Megatron-LM/pull/5700/commits/ce43b4fb06b08932dfceb529de22c69d20120ebc)

**Comment by guihong-nv** [comment]
/ok to test [f9946f5](https://github.com/NVIDIA/Megatron-LM/pull/5700/commits/f9946f50d958728ea1a37ba22d3d6ee078e8a853)

**Comment by guihong-nv** [comment]
/ok to test [8a96c6b](https://github.com/NVIDIA/Megatron-LM/pull/5700/commits/8a96c6b7bf263fadc7520fe10b9822a8cde92f48)

**Comment by guihong-nv** [comment]
/ok to test [2d8a3f0](https://github.com/NVIDIA/Megatron-LM/pull/5700/commits/2d8a3f011a5993d8d1bf1b651f1da8453529495b)

**Comment by shanmugamr1992** [comment]
/ok to test [1524504](https://github.com/NVIDIA/Megatron-LM/pull/5700/commits/15245041e771c84820438b99cd8bc82ebf7ec594)

**Comment by shanmugamr1992** [comment]
### NeMo-RL functional tests

megatron-lm: `15245041` · megatron-bridge: `a056408d` · suite: `L1` · cluster: `oci-hsg` · updated 2026-07-27 20:17 UTC

| Test | Status | Comment |
| --- | --- | --- |
| `grpo_megatron_generation` | 🛠️ fixed | Was failing on `prompt_tokens=None`; not caused by this PR (reproduces on megatron-core `main`). Fixed by NVIDIA-NeMo/RL#3363 (draft, needs review). Passes with it applied. |
| `grpo_megatron_generation_non_colocated` | 🛠️ fixed | Same `prompt_tokens=None` cause; fixed by NVIDIA-NeMo/RL#3363 (draft). Passes with it applied. |
| `grpo_megatron_generation_async` | 🛠️ fixed | Previously exited 0 while EVERY generation raised the same `prompt_tokens=None` error, because the rollout loop catches per-sample generation exceptions. It was training on empty generations. Now genuinely passes with NVIDIA-NeMo/RL#3363 (draft). |
| `grpo_megatron_generation_colocated_async` | 🛠️ fixed | Also a silent pass before, for the same reason. Now genuinely passes with NVIDIA-NeMo/RL#3363 (draft). |
| `grpo_megatron_generation_async_gym` | ✅ pass |  |
| `grpo_megatron_generation_multiturn` | ⚠️ fail (pre-existing) | No longer crashes with the fix applied, but fails a numerics check: median(train/token_mult_prob_error) = 1.99 against a threshold of 1.1, with a 100% truncation rate. Confirmed identical on megatron-core `main` (076f61fc), so it is not caused by this PR. It is a separate pre-existing problem that the crash was masking; tracked on the NeMo-RL side. |

**Nothing here is caused by this PR.** Every failure reproduces identically on
megatron-core `main` (`076f61fc`) with the same Megatron-Bridge pin.

Four of these tests were broken by megatron-core
[#5918](https://github.com/NVIDIA/Megatron-LM/pull/5918), which stopped the engine
echoing `prompt_tokens` back unless `SamplingParams.return_prompt_tokens` is set.
NeMo-RL was reading the prompt out of the engine's reply and got `None`. Two of the
four were failing *silently* — they exited 0 while every generation raised, because
NeMo-RL's rollout loop catches per-sample generation errors, so they trained on
empty output and reported success.

Fix for review: **[NVIDIA-NeMo/RL#3363](https://github.com/NVIDIA-NeMo/RL/pull/3363)**
(draft). Nothing is needed on this PR for it.

`grpo_megatron_generation_multiturn` remains red for an unrelated, pre-existing
numerics reason that the crash had been hiding. It is being tracked separately on
the NeMo-RL side.

<sub>Posted by the nemo-rl-testing-agent. Re-runs edit this comment in place.</sub>

**Comment by guihong-nv** [comment]
/ok to test [dccfee3](https://github.com/NVIDIA/Megatron-LM/pull/5700/commits/dccfee394c45ec90b9a5557dac34ecf489083c0a)

**Comment by shanmugamr1992** [comment]
/ok to test [e62bc42](https://github.com/NVIDIA/Megatron-LM/pull/5700/commits/e62bc42d3dba2b5b3a8f5a80547478d9ef6123af)

**Comment by shanmugamr1992** [comment]
/ok to test [864f01e](https://github.com/NVIDIA/Megatron-LM/pull/5700/commits/864f01e73d60b9c2242cfe602c613f321fd942fe)

**Review by wdykas — APPROVED** [review]

**Comment by shanmugamr1992** [comment]
/ok to test [9bd87f5](https://github.com/NVIDIA/Megatron-LM/pull/5700/commits/9bd87f5f6feb06596e2f4422fdd10f634a3721e2)
